framework/src/interface/packet.rs

```diff
@@ -347,6 +379,34 @@ impl<T: EndOffset, M: Sized + Send> Packet<T, M> {
         }
     }
 
+    #[inline]
+    pub fn insert_header<T2: EndOffset<PreviousHeader = T>>(&mut self, header: &T2) -> Result<()> {
+        unsafe {
+            let len = self.data_len();
+            let size = header.offset();
+            let added = (*self.mbuf).add_data_end(size);
+
+            let hdr = header as *const T2;
+            let offset = self.offset() + self.payload_offset();
```

Comment:
so what exactly do we expect to come back from this Packet::offset() ? not totally sure what `offset` winds up being here tbh

Reply:
Yeah the naming on these things are great. Some offsets are absolute while others are relative etc. Let me see if I can get this right.

So the entire network packet is stored in a mbuf by DPDK. NetBricks packet is a wrapper with an offset that tells you where this particular header starts in the mbuf. If you have a `Packet<TcpHeader>`, then `self.offset()` is where the first byte of the tcp header is. `self.payload_offset()` is actually just `header.size()`. I know :headbang:. `self.offset() + self.payload_offset()` gives you where the tcp payload starts in this example.

Also, this is a copy-pasta job from `push_header`. Can't totally blame @zeeshanlakhani for the gnarliness.

Reply:
Yeah maybe some of the variable names could be better, but I think @drunkirishcoder has the right intuition about how it works.

Reply:
Yep. comments added, a bit better naming, but not much to change it throughout right now. But, I added nicer comments, including Dan's intuition.